openssl-dynamic/src/main/c/sslcontext.c

```diff
@@ -2331,6 +2342,141 @@ TCN_IMPLEMENT_CALL(void, SSLContext, setPrivateKeyMethod)(TCN_STDARGS, jlong ctx
 #endif // OPENSSL_IS_BORINGSSL
 }
 
+static int tcn_new_session_cb(SSL *ssl, SSL_SESSION *session) {
+    JNIEnv *e = NULL;
+    jboolean result = JNI_FALSE;
+    tcn_ssl_ctxt_t *c = NULL;
+
+    TCN_GET_SSL_CTX(ssl, c);
+    TCN_ASSERT(c != NULL);
+
+    if (tcn_get_java_env(&e) != JNI_OK) {
+        return 0;
+    }
+    if (c->ssl_session_cache == NULL) {
+        return 0;
+    }
+
+    result = (*e)->CallBooleanMethod(e, c->ssl_session_cache, c->ssl_session_cache_creation_method, P2J(ssl), P2J(session));
+
+    if ((*e)->ExceptionCheck(e)) {
+        return 0;
+    }
+
+    if (result == JNI_TRUE) {
+        return 1;
+    }
+    return 0;
+}
+
+static void tcn_remove_session_cb(SSL_CTX *ctx, SSL_SESSION *session) {
+    JNIEnv *e = NULL;
+    tcn_ssl_ctxt_t *c = tcn_SSL_CTX_get_app_state(ctx);
+
+    TCN_ASSERT(c != NULL);
+
+    if (tcn_get_java_env(&e) != JNI_OK) {
+        return;
+    }
+    if (c->ssl_session_cache == NULL) {
+        return;
+    }
+
+    (*e)->CallVoidMethod(e, c->ssl_session_cache, c->ssl_session_cache_removal_method, P2J(ctx), P2J(session));
+}
+
+static SSL_SESSION* tcn_get_session_cb(SSL *ssl, const unsigned char *session_id, int len, int *copy) {
+    JNIEnv *e = NULL;
+    jlong result = JNI_FALSE;
+    tcn_ssl_ctxt_t *c = NULL;
+    jbyteArray bArray = NULL;
+
+    TCN_GET_SSL_CTX(ssl, c);
+    TCN_ASSERT(c != NULL);
+
+    if (tcn_get_java_env(&e) != JNI_OK) {
+        return 0;
```

Comment:
`return NULL;`? (not too big a deal) Same with the one right below.

Reply:
good catch